src/components/views/voice_messages/Waveform.tsx

```diff
@@ -28,17 +30,30 @@ interface IState {
  * A simple waveform component. This renders bars (centered vertically) for each
  * height provided in the component properties. Updating the properties will update
  * the rendered waveform.
+ *
+ * For CSS purposes, a mx_Waveform_bar_100pct class is added when the bar should be
+ * "filled", as a demonstration of the progress property.
  */
 @replaceableComponent("views.voice_messages.Waveform")
 export default class Waveform extends React.PureComponent<IProps, IState> {
+    public static defaultProps = {
+        progress: 1,
+    };
+
     public constructor(props) {
         super(props);
     }
 
     public render() {
         return <div className='mx_Waveform'>
             {this.props.relHeights.map((h, i) => {
-                return <span key={i} style={{height: (h * 100) + '%'}} className='mx_Waveform_bar' />;
+                const progress = this.props.progress;
+                const isCompleteBar = (i / this.props.relHeights.length) <= progress && progress > 0;
+                const classes = classNames({
+                    'mx_Waveform_bar': true,
+                    'mx_Waveform_bar_100pct': isCompleteBar,
+                });
+                return <span key={i} style={{height: (h * 100) + '%'}} className={classes} />;
```

Comment:
I found that this kind of rapid changes on the `height` property can actually cost quite a bit as it will [trigger layout/paint/compositing](https://csstriggers.com/height).

However you can skip the layout and paint operaiton if all of your span have a `height: 100%` and you play with [`transform: scaleY(percentage)`](https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/scaleY()), Blink and Gecko are [optimising those operations](https://csstriggers.com/transform)

Reply:
Will look into it for a future change